File: providers/glm.py
"""
智谱 GLM Coding Plan 配额查询适配器

智谱开放平台没有公开的预付费余额接口，这里查的是 Coding Plan 套餐配额。
接口按时间窗口返回若干条限额（5 小时 / 周 / 月，类型有 TOKENS_LIMIT / TIME_LIMIT /
CREDIT_LIMIT 几种），这里取「剩余比例最低」的那个窗口作为余额，单位是百分比：
100 表示一点没用，0 表示用光。阈值也按百分比填，例如 10 表示剩余不足 10% 时告警。

Z.ai 国际站是同一套接口，host 换成 https://api.z.ai 即可。
"""
from .base import ProviderSpec, SimpleHTTPProvider


def _check(data):
    if not data.get('success') or data.get('code') not in (None, 200):
        return f"API 返回错误: {data.get('msg', '未知错误')}"
    return None


def _remaining_percent(limit):
    """单个窗口的剩余比例。有总量/剩余量就精确算，否则用接口给的已用百分比反推。"""
    usage, remaining = limit.get('usage'), limit.get('remaining')
    if isinstance(usage, (int, float)) and usage > 0 and isinstance(remaining, (int, float)):
        return max(0.0, min(100.0, float(remaining) / usage * 100))

    percentage = limit.get('percentage')
    if isinstance(percentage, (int, float)):
        return max(0.0, min(100.0, 100.0 - percentage))
    return None


def _extract(data):
    limits = (data.get('data') or {}).get('limits')
    if not isinstance(limits, list):
        raise ValueError("无法从响应中解析 data.limits 字段")

    values = [v for v in (_remaining_percent(l) for l in limits if isinstance(l, dict)) if v is not None]
    if not values:
        raise ValueError("data.limits 里没有可解析的配额窗口")
    return round(min(values), 2)


class GLMProvider(SimpleHTTPProvider):
    """智谱 GLM Coding Plan 适配器"""

    SPEC = ProviderSpec(
        name="GLM",
        url="https://open.bigmodel.cn/api/monitor/usage/quota/limit",
        headers={"Accept": "application/json"},
        check=_check,
        extract=_extract,
    )

File: internal/provider/glm.go
package provider

import (
	"errors"
	"fmt"
	"math"
)

// Implementation note.
// Implementation note.
// Implementation note.
// Implementation note.
//
// Implementation note.
var glmSpec = Spec{
	Key:         "glm",
	Name:        "GLM",
	DefaultType: "quota",
	URL:         "https://open.bigmodel.cn/api/monitor/usage/quota/limit",
	Headers:     map[string]string{"Accept": "application/json"},
	Check: func(data map[string]any) error {
		if !truthy(data["success"]) || !glmCodeOK(data["code"]) {
			return fmt.Errorf("API returned an error: %s", messageOr(data, "msg", "Unknown error"))
		}
		return nil
	},
	Extract: func(data map[string]any) (float64, error) {
		limits, ok := Dig(data, "data", "limits").([]any)
		if !ok {
			return 0, errors.New("Could not parse data.limits field")
		}

		// Implementation note.
		lowest := math.Inf(1)
		for _, item := range limits {
			limit := Object(item)
			if limit == nil {
				continue
			}
			if percent, ok := glmRemainingPercent(limit); ok && percent < lowest {
				lowest = percent
			}
		}
		if math.IsInf(lowest, 1) {
			return 0, errors.New("data.limits contains no parseable quota window")
		}
		return round2(lowest), nil
	},
}

// Implementation note.
// Implementation note.
// Implementation note.
func glmCodeOK(code any) bool {
	if code == nil {
		return true
	}
	value, ok := jsonNum(code)
	return ok && value == 200
}

// Implementation note.
// Implementation note.
func glmRemainingPercent(limit map[string]any) (float64, bool) {
	usage, hasUsage := jsonNum(limit["usage"])
	remaining, hasRemaining := jsonNum(limit["remaining"])
	if hasUsage && usage > 0 && hasRemaining {
		return clampPercent(remaining / usage * 100), true
	}
	if percentage, ok := jsonNum(limit["percentage"]); ok {
		return clampPercent(100 - percentage), true
	}
	return 0, false
}

func clampPercent(v float64) float64 {
	return math.Max(0, math.Min(100, v))
}

func init() { RegisterSpec(glmSpec) }
